### DISCUSSION TITLE
How to catch 404 in layout.js and middleware.js

### DISCUSSION THREAD

**Discussion by MasterProvider** [vercel/next.js#59576] [Help]
### Summary

![Screenshot 2023-12-13 221954](https://github.com/vercel/next.js/assets/48547307/c623e9d5-aaf6-4b0b-ac62-e479a3f254f7)
how to detect path is 404 not found

\middleware.js

if(isNotfound){
  console.log("page not found")
}

\app\layout.js

if(isNotfound){
  console.log("page not found")
}

**Comment by Sirivasv** [comment]
For the middleware.js part according to the [docs](https://nextjs.org/docs/app/building-your-application/routing/middleware):
> Middleware runs before cached content and routes are matched.

So in the middleware.js you define the `isNotfound` yourself either by having matchers for the known routes and then a default one which can act as your "Not found" for the rest. Like in this example in the same docs page:

```
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    '/((?!api|_next/static|_next/image|favicon.ico).*)',
  ],
}
```

And for the layout as I understand it seems like the opposite it is executed once that the route is matched [source](https://nextjs.org/docs/pages/building-your-application/routing/pages-and-layouts).

But maybe you can also define a custom 404 so that each time a 404 is returned by your app you execute some code:
https://nextjs.org/docs/pages/building-your-application/routing/custom-error#404-page
